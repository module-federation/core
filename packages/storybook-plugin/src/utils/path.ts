import os from 'os';
import path from 'path';

export const correctImportPath = (context: string, entryFile: string) => {
  if (os.platform() !== 'win32') {
    return entryFile;
  }

  if (entryFile.match(/^\.?\.\\/) || !entryFile.match(/^[A-Z]:\\\\/i)) {
    return entryFile.replace(/\\/g, '/');
  }

  const joint = path.win32.relative(context, entryFile);
  const relative = joint.replace(/\\/g, '/');

  if (relative.includes('node_modules/')) {
    return relative.split('node_modules/')[1];
  }

  return `./${relative}`;
};
