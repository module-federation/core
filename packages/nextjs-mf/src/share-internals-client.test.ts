import {
  getNextInternalsShareScopeClient,
  getPagesDirSharesClient,
  getAppDirSharesClient,
} from './share-internals-client';
import fs from 'fs';
import os from 'os';
import path from 'path';

// Mock getReactVersionSafely from ./internal-helpers
jest.mock('./internal-helpers', () => ({
  getReactVersionSafely: jest.fn(() => '18.2.0'),
  // Provide a safeRequireResolve mock that behaves like require.resolve with fallback
  safeRequireResolve: jest.fn((id: string, options?: any) => {
    try {
      return require.resolve(id, options);
    } catch {
      return id; // fall back to original id if not resolvable in test
    }
  }),
}));

describe('getNextInternalsShareScopeClient', () => {
  let tempDir: string;

  // Helper to create a temp dir with a package.json for next
  const setupNextVersion = (version: string) => {
    tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'nextjs-mf-test-'));
    const pkgPath = path.join(tempDir, 'node_modules', 'next', 'package.json');
    fs.mkdirSync(path.dirname(pkgPath), { recursive: true });
    fs.writeFileSync(pkgPath, JSON.stringify({ version }), 'utf8');
    return tempDir;
  };

  afterEach(() => {
    if (tempDir && fs.existsSync(tempDir)) {
      fs.rmSync(tempDir, { recursive: true, force: true });
    }
    jest.resetModules();
    jest.clearAllMocks();
  });

  it('returns the correct share scope for a client compiler (Next 15)', () => {
    const context = setupNextVersion('15.0.0');
    const compiler = {
      options: { name: 'client' },
      context,
    } as any;
    const result = getNextInternalsShareScopeClient(compiler);
    expect(result).toMatchSnapshot();
  });

  it('returns the correct share scope for a client compiler (Next 14)', () => {
    const context = setupNextVersion('14.0.0');
    const compiler = {
      options: { name: 'client' },
      context,
    } as any;
    const result = getNextInternalsShareScopeClient(compiler);
    expect(result).toMatchSnapshot();
  });

  it('returns an empty object for a non-client compiler', () => {
    const context = setupNextVersion('15.0.0');
    const compiler = {
      options: { name: 'server' },
      context,
    } as any;
    const result = getNextInternalsShareScopeClient(compiler);
    expect(result).toEqual({});
  });

  // Tests specifically for getPagesDirSharesClient
  describe('getPagesDirSharesClient', () => {
    it('returns the correct config for Next 15', () => {
      const context = setupNextVersion('15.0.0');
      const compiler = { context } as any;
      const result = getPagesDirSharesClient(compiler);
      expect(result).toMatchSnapshot();
    });

    it('returns the correct config for Next 14', () => {
      const context = setupNextVersion('14.0.0');
      const compiler = { context } as any;
      const result = getPagesDirSharesClient(compiler);
      expect(result).toMatchSnapshot();
    });
  });

  // Tests specifically for getAppDirSharesClient
  describe('getAppDirSharesClient', () => {
    it('returns the correct config for Next 15', () => {
      const context = setupNextVersion('15.0.0');
      const compiler = { context } as any;
      const result = getAppDirSharesClient(compiler);
      expect(result).toMatchSnapshot();
    });

    it('returns the correct config for Next 14', () => {
      const context = setupNextVersion('14.0.0');
      const compiler = { context } as any;
      const result = getAppDirSharesClient(compiler);
      expect(result).toMatchSnapshot();
    });
  });
});
