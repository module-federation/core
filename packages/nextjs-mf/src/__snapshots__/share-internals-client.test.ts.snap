// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getNextInternalsShareScopeClient returns the correct share scope for a client compiler (Next 14) 1`] = `
Object {
  "next-compiled": Object {
    "issuerLayer": "app-pages-browser",
    "layer": "app-pages-browser",
    "request": "next/dist/compiled/",
    "shareKey": "next/dist/compiled/",
    "shareScope": "app-pages-browser",
    "singleton": true,
  },
  "next-compiled-react": Object {
    "exclude": Object {
      "request": /\\^\\(\\?!\\.\\*react\\)\\.\\*\\|react-refresh/,
    },
    "issuerLayer": "app-pages-browser",
    "layer": "app-pages-browser",
    "request": "next/dist/compiled/",
    "requiredVersion": "^18.2.0",
    "shareKey": "next/dist/compiled/",
    "shareScope": "app-pages-browser",
    "singleton": true,
    "version": "18.2.0",
  },
  "next/": Object {
    "exclude": Object {
      "request": /dist/,
    },
    "singleton": true,
  },
  "react-dom-original": Object {
    "request": "react-dom",
    "shareKey": "react-dom",
    "shareScope": "default",
    "singleton": true,
  },
  "react-dom/client-original": Object {
    "request": "react-dom/client",
    "shareKey": "react-dom/client",
    "shareScope": "default",
    "singleton": true,
  },
  "react-original": Object {
    "request": "react",
    "shareKey": "react",
    "shareScope": "default",
    "singleton": true,
  },
  "react/jsx-dev-runtime-original": Object {
    "request": "react/jsx-dev-runtime",
    "shareKey": "react/jsx-dev-runtime",
    "shareScope": "default",
    "singleton": true,
  },
  "react/jsx-runtime-original": Object {
    "request": "react/jsx-runtime",
    "shareKey": "react/jsx-runtime",
    "shareScope": "default",
    "singleton": true,
  },
  "styled-jsx": Object {
    "singleton": true,
  },
  "styled-jsx/": Object {
    "singleton": true,
  },
}
`;

exports[`getNextInternalsShareScopeClient returns the correct share scope for a client compiler (Next 15) 1`] = `
Object {
  "next-compiled": Object {
    "issuerLayer": "app-pages-browser",
    "layer": "app-pages-browser",
    "request": "next/dist/compiled/",
    "shareKey": "next/dist/compiled/",
    "shareScope": "app-pages-browser",
    "singleton": true,
  },
  "next-compiled-react": Object {
    "exclude": Object {
      "request": /\\^\\(\\?!\\.\\*react\\)\\.\\*\\|react-refresh/,
    },
    "issuerLayer": "app-pages-browser",
    "layer": "app-pages-browser",
    "request": "next/dist/compiled/",
    "requiredVersion": "^18.2.0",
    "shareKey": "next/dist/compiled/",
    "shareScope": "app-pages-browser",
    "singleton": true,
    "version": "18.2.0",
  },
  "next/": Object {
    "exclude": Object {
      "request": /dist/,
    },
    "singleton": true,
  },
  "react-dom-original": Object {
    "exclude": Object {
      "version": "18.x",
    },
    "request": "react-dom",
    "shareKey": "react-dom",
    "shareScope": "default",
    "singleton": true,
  },
  "react-dom/client-original": Object {
    "exclude": Object {
      "version": "18.x",
    },
    "request": "react-dom/client",
    "shareKey": "react-dom/client",
    "shareScope": "default",
    "singleton": true,
  },
  "react-original": Object {
    "exclude": Object {
      "version": "18.x",
    },
    "request": "react",
    "shareKey": "react",
    "shareScope": "default",
    "singleton": true,
  },
  "react/jsx-dev-runtime-original": Object {
    "exclude": Object {
      "version": "18.x",
    },
    "request": "react/jsx-dev-runtime",
    "shareKey": "react/jsx-dev-runtime",
    "shareScope": "default",
    "singleton": true,
  },
  "react/jsx-runtime-original": Object {
    "request": "react/jsx-runtime",
    "shareKey": "react/jsx-runtime",
    "shareScope": "default",
    "singleton": true,
  },
  "styled-jsx": Object {
    "singleton": true,
  },
  "styled-jsx/": Object {
    "singleton": true,
  },
}
`;
